### ISSUE TITLE
Sub to trigger many subs

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#158]
@jordanh I'm going down a biiiig rabbit hole here, wanted to get your thoughts:

- As discussed, we're gonna subscribe to `teamMembers/:teamId`. That will return `t` documents, 1 for each `teamMember`.
- Then, we need to subscribe to all of their projects, so we call `projects/:teamMemberId` `t` times.
```
const makeProjectSubs = (teamMembers) => {
  const projectSubs = [];
  for (let i = 0; i < teamMembers.length; i++) {
    const teamMemberId = teamMembers[i].id;
    projectSubs[i] = cashay.subscribe(projectSubString, subscriber, {
      component: 'projectSub',
      key: teamMemberId,
      variables: {teamMemberId}
    });
  }
  return projectSubs;
};
```
- When i perform this inside `mapStateToProps` I get the classic tenderfoot error of `warning.js:44 Warning: setState(...): Cannot update during an existing state transition`. 
- This is what you get when you call `dispatch` in a `render` method. Why does it happen here?
 - Well, calling `cashay.subscribe` initiates the `subscriber`
 - The `subscriber` calls `socket.subscribe`
 - `socket.subscribe` fires off a `subscribeRequest` event
 - the redux-socket-cluster HOC has a `subscribeRequest` event listener that calls `dispatch` to put the channel into the `pendingSubs` array of the redux store
 - Technically, we're in the `connect#render`function (connect comes from react-redux), which I'm guessing is why it's throwing the warning. Why it _only_ throws when I run a loop is beyond me. If I do them one-by-one, it doesn't throw. As I see it, _it should probably throw of every subscribe_.

So, I see a few options:
- Delay the socketRequest event by patching socketCluster-client
- wrap the `dispatch` in a `setTimeout` in redux-socket-cluster (although any other subscribeRequest listeners (of which we have none) would bork)
- Do nothing & hope react fixes this?

I'm gonna take option 2 just to keep moving forward, but from past experience, when I need to use `setTimeout(() => {},0)`, it usually means something is very wrong. I've been down in the weeds for far too long on this. From what you see, are there any cleaner alternatives?

**Comment by jordanh** [comment]
Wow.

This made no sense to me until you sent me down the rabbit hole with you and I started reading the source more carefully.

I see now that `socketCluster-client` is using a Component emitter pattern to propagate the `subscribeRequest` event. That event propagation is actually synchronous when emit() is being called ([see here](https://github.com/component/emitter/blob/master/index.js#L125), sneaky!). So, yeah, when `redux-socket-cluster`'s `subscribeRequest` event listener is called it will call `dispatch`within `connect#render`. That stinks. (Why you don't see those warnings outside running them in a loop is currently beyond me. But let's set that aside for now.)

I was curious how `cashay.query()` got around this same limitation. After all, isn't it a call that originates in mSTP which will eventually result in calls to `dispatch`?

When I read the source, I saw that – ah! – `cashay.query` [will eventually call queryServer](https://github.com/mattkrick/cashay/blob/master/src/Cashay.js#L267) which is async! So `cashay.query` first returns the shape of the data by using the AST, the React render loop completes, then the server is queried and the redux action is dispatched. When mSTP comes around again, it's cached in the store. Clever!

Could cashay.subscribe be refactored the same way? If it returns the same data shape as `cashay.query` but propagated the `subscribeRequest` event as a promise, wouldn't that fix it?

**Comment by mattkrick** [comment]
heh... yeah, the `Emitter` is a sneaky fella. Another post-mvp todo item is to rewrite the socketcluster-client in ES2015 & take advantage of things like WeakSets for storing the watchers. Should cut memory use & the API complexity in half, but that's after we get some good CmRDT going strong.

good idea about refactoring cashay.subscribe!
implemented in the newest cashay, works like a champ.
Now, when you start a subscription, the status is `subscribing`.
When the status is `subscribing` then `unsubscribe === null`.
Then on the next tick, `status === 'subscribed` and `unsubscribe === function()`

**Comment by jordanh** [comment]
I think this is a rad design.

**Comment by mattkrick** [comment]
live in cashay v0.15.0.
will close at end of team-dash sprint

**Comment by mattkrick** [comment]
fixed by #181
